### PR TITLE
Speedup test string substitutions from .source to .sql/.out files

### PR DESCRIPTION
**Pull request by ashwinstar**
In GPDB, gpstringsubs.pl is used to replace tokens in .source test
files when creating .sql/.out files for pg_regress. pg_regress itself
also performs the substitutions. The order is pg_regress performs
substitutions and after that remaining ones are performed by
gpstringsubs.pl if required.

The logic to check if there are still tokens remaining to be replaced
after pg_regress has performed replacement was based on checking if
'@' character is present in file or not. If '@' is present in file
then gpstringsubs.pl is invoked for that file. gpstringsubs.pl is
super slow and takes long time. In many .source files keywords exist
like `@Description` or `@db_name` (specially for UAO related tests),
which don't need any substitution. But due to presence of these '@'
charater keywords, gpstringsubs.pl was invoked for such files
unnecessarily, causing the slow down.

So, modifying the check to invoke gpstringsubs.pl only on presence of
"@gp" keyword instead. So, now any token replacements handled by
gpstringsubs.pl should start with "@gp" in the name. Hence, renamed
`@syslocale@` to "@gp_syslocale@".

Also, moved the logic to perform replacement for `@hostname@` and
`@gpcurusername@` to pg_regress itself, as it's much faster and easier
to do the same in C.

Given the whole testsuite takes very long time in GPD. Need to iterate
on single test routinely. These substitutions are invoked on every run
of pg_regress (mainly isolation2 also), hence helps to reduce the
iteration time for running single test in test suite during
development, along with helping cut down pg_regress file conversion
time in CI. For regress directory cuts down time from 8 secs and
isolation2 from 25 secs to few msecs.

In long run we should complete move all the logic in gpstringsubs.pl
to pg_regress and kill gpstringsubs.pl.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/test/regress/pg_regress.c [hunk 1]
```diff
@@ -3188,3 +3201,54 @@ regression_main(int argc, char *argv[], init_function ifunc, test_function tfunc
...
+static char *
+get_host_name(int16 contentid, char role)
+{
+	char psql_cmd[MAXPGPATH];
+	FILE       *fp;
+	char line[1024];
+	int len;
+	char *hostname = NULL;
+
+	len = snprintf(psql_cmd, sizeof(psql_cmd),
```
Comment (asimrp) [inline src/test/regress/pg_regress.c]: FTS now writes the configuration to flat file named `gpsegconfig_dump`.  Reading hostname from that file is also an option.

It seems that this hostname token appears only in `location` clause of `create external table` command within the tests.  There are only a few tests using it.  A better option is to generate `create external table` command from within the test case, say, by writing a UDF that forms the create table SQL statement and executes it.  Burdening the test harness for such a specific use case doesn't seem right.  Of course, this suggestion is not applicable to this PR but may be follow up.

**Comment by ashwinstar** [comment]
On Wed, Sep 4, 2019 at 9:59 PM Asim R P <notifications@github.com> wrote:

> *@asimrp* approved this pull request.
>
> Several seconds to a few milliseconds is great improvement. I've been so
> annoyed by the long time it takes to run a single isolation2 test. Thank
> you for fixing that!
> ------------------------------
>
> In src/test/regress/pg_regress.c
> <https://github.com/greenplum-db/gpdb/pull/8559#discussion_r321073891>:
>
> > @@ -3188,3 +3201,54 @@ regression_main(int argc, char *argv[], init_function ifunc, test_function tfunc
>
>  	return 0;
>  }
> +
> +static char *
> +get_host_name(int16 contentid, char role)
> +{
> +	char psql_cmd[MAXPGPATH];
> +	FILE       *fp;
> +	char line[1024];
> +	int len;
> +	char *hostname = NULL;
> +
> +	len = snprintf(psql_cmd, sizeof(psql_cmd),
>
> FTS now writes the configuration to flat file named gpsegconfig_dump.
> Reading hostname from that file is also an option.
>
Thanks for the review Asim! I would like to keep usage of that file
internal to FTS/dispatch only and not really like any other code to use it.
That way it abstracts it to FTS and can easily be modified in the future if
needed. Unnecessarily increasing dependency on that file, where not
required seems a violation of layering/abstraction.

> It seems that this hostname token appears only in location clause of create
> external table command within the tests. There are only a few tests using
> it. A better option is to generate create external table command from
> within the test case, say, by writing a UDF that forms the create table SQL
> statement and executes it. Burdening the test harness for such a specific
> use case doesn't seem right. Of course, this suggestion is not applicable
> to this PR but may be follow up.
>
`@hostname@` is used in many tests like in contrib, gpcontrib, src/bin, ...
directories, so seem will need something generic anyways. I am not sure
though why all such test relies on content zero's hostname only though,
might be master's hostname is fine as well. I lazily ignored to dig into
that detail.

**Comment by ashwinstar** [comment]
Merged c8907e821a53d82754927d5e678dfa6afc0ad148

**Comment by danielgustafsson** [comment]
Meh .. this got merged while I looked at it, unfortunate timing

**Comment by ashwinstar** [comment]
On Thu, Sep 5, 2019 at 11:06 AM Daniel Gustafsson <notifications@github.com>
wrote:

> Meh .. this got merged while I looked at it, unfortunate timing
>

Sorry. Please provide all the comments, I will patch it up and commit by
incorporating the changes.

File: src/test/regress/pg_regress.c [hunk 2]
```diff
@@ -3188,3 +3201,54 @@ regression_main(int argc, char *argv[], init_function ifunc, test_function tfunc
...
+
+	if (pclose(fp) < 0)
+	{
+		fprintf(stderr, "%s: cannot close shell command\n", progname);
+		exit_nicely(2);
+	}
+
+	hostname = psprintf("%s", trim_white_space(line));
+
+	if (hostname == NULL)
```
Comment (danielgustafsson) [inline src/test/regress/pg_regress.c]: `psprintf()` wont return NULL on error.

**Comment by danielgustafsson** [comment]
n 5 Sep 2019, at 20:21, Ashwin Agrawal <notifications@github.com> wrote:
> 
> On Thu, Sep 5, 2019 at 11:06 AM Daniel Gustafsson <notifications@github.com>
> wrote:
> 
> > Meh .. this got merged while I looked at it, unfortunate timing
> >
> 
> Sorry.

Not at all, it was me being slow.

> Please provide all the comments, I will patch it up and commit by
> incorporating the changes.

I’ve added all my comments, nothing terribly exciting but a few small things spotted.

**Pull request closed without merging**